contracts/coconsensus/Coconsensus.sol

```diff
@@ -46,23 +46,171 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
 
     /**
      * Mapping of metachain id to latest block number(tip) stored
      * in blockchains.
      */
-    mapping (bytes32 /* metachainId */ => uint256 /* blocknumber */) blockTips;
+    mapping (bytes32 /* metachainId */ => uint256 /* blocknumber */) public blockTips;
 
     /** Mapping of metachain id to the protocore contract address. */
-    mapping (bytes32 /* metachainId */ => ProtocoreI) protocores;
+    mapping (bytes32 /* metachainId */ => ProtocoreI) public protocores;
 
     /** Mapping of metachain id to the observers contract address. */
-    mapping (bytes32 /* metachainId */ => ObserverI) observers;
+    mapping (bytes32 /* metachainId */ => ObserverI) public observers;
 
     /** Mapping of metachain id to the domain separators. */
-    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) domainSeparators;
+    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) public domainSeparators;
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Setup function does the initialization of all the mosaic
+     *         contracts on the auxiliary chain.
+     *
+     * @dev This function can be called only once.
+     */
+    function setup() public {
+
+        require(
+            originMetachainId == bytes32(0),
+            "Coconsensus contract is already initialized."
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+        auxiliaryMetachainId = genesisAuxiliaryMetachainId;
+
+        /*
+         * Setup self protocore contract first. All other protocores will
+         * require the dynasty from the self protocore to finalise the genesis
+         * checkpoint.
+         */
+        setupProtocores(auxiliaryMetachainId);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+            /*
+             * The setup of self protocore is already done as a first step so
+             * skip the protocore setup if the current metachain id is equal to
+             * the auxiliary metachain id.
+             */
+            if (currentMetachainId != auxiliaryMetachainId) {
+                // Setup protocore contract for the given metachain id.
+                setupProtocores(currentMetachainId);
+            }
+
+            // Setup observer contract for the given metachain id.
+            setupObservers(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* Private Functions */
+
+    /**
+     * @notice Do the initial setup of protocore contract and initialize the
+     *         storage of coconsensus contract.
+     *
+     * @param _metachainId Metachain id.
+     *
+     * /pre `protocoreAddress` is not 0
+     *
+     * /post Setup protocore contract.
```

Comment:
Suggestion:

\post condition documentations can be improved.

For example:
    * \post Adds newly setup protocore's domain separator into `domainSeparators` storage variable.
    * \post Updates `blockTips` storage variable with the latest finalized checkpoint's block number of the newly setup protocore.

and so on

Reply:
Updated, please check.